src/Data/DecisionDiagram/BDD.hs

```diff
@@ -391,8 +390,9 @@ pbAtLeast xs k0 = unfoldOrd f (0, k0)
 
     f :: (Int, w) -> Sig (Int, w)
     f (!i, !k)
-      | not (k <= ub) = SLeaf False
+      | k > ub = SLeaf False
       | i == V.length xs' && 0 >= k = SLeaf True
+      -- !!! SUS?
```

Comment:
The function `f` encodes inequality constraint `wi*xi + … + wn*xn ≥ k`.

Here `ub := max_{xi,…,xn} wi*xi + … + wn*xn` and `lb := min_{xi,…,xn} wi*xi + … + wn*xn` .

* In the clause above the comment, the condition `i == V.length xs' && 0 >= k` means that the left-hand side `wi*xi + … + wn*xn` is empty and the right-hand side `k` is less than or equal to 0. Therefore `wi*xi + … + wn*xn ≥ k` is true.

* In the clause below the comment, the condition `lb ≥ k` implies`wi*xi + … + wn*xn ≥ lb ≥ k`. Therefore `wi*xi + … + wn*xn ≥ k` is also true.

Now I realized that the former condition `i == V.length xs' && 0 >= k` implies the latter condition `lb ≥ k`, so the former clause is redundant. Similarly, `pbAtMost` function below also has the redundant clause `i == V.length xs' && 0 <= k = SLeaf True`.

If this explains what you suspected, please add a commit to delete those two clauses (`      | i == V.length xs' && 0 >= k = SLeaf True` line above and `      | i == V.length xs' && 0 <= k = SLeaf True` line in the below `pbAtMost`) and the comment line `      -- !!! SUS?`.

Reply:
Returning to it, I think the "issue" I saw was different (I thought there was difference between equivalent cases of BDD and ZDD, but I checked again and they were the same.

However I looked into what you wrote and it makes sense/is true. To double check I tried running tests with commented out those two clauses. So I'll obliterate the comment and clauses and clean up the commits.